### PR TITLE
Midproject

### PR DESCRIPTION
**Pull request by brianrice2**
Hello CH, hope you're doing well. This PR contains substantial changes so let me give you a quick walkthrough:

**Summary of changes:**

- Add Github Actions for automatic table of contents refreshing (`.github/workflows`)
- Try to set up a GA for automatic Sphinx documentation rebuilding. This didn't work and cluttered up the commit history, so in retrospect would have been a _much_ better idea to create a branch for this purpose.
  - Documentation is refreshed manually currently
- Remove files from the template that aren't necessary for the current stage of the project. They may be added in later.
- Add `src/` modules for data acquisition and relational database setup:
  - `load_data` downloads the raw data (static file and no login/paywall required) and uploads to/downloads from S3. A copy of the raw data is saved in `data/raw/` purely for convenience.
  - `add_albums` creates, deletes, and populates a MySQL/SQLite database
- Unit tests for an S3 path parsing function. Most of the work involves database queries or network calls, so these do not have associated tests per the implementation guidelines from lecture 3.

**Feedback requested:**

- Logging has appropriate messages and levels with sufficient coverage
- Sufficient error handling
- Additional documentation would be helpful for getting the user up to speed
- Anything else you feel worth mentioning (for example, unused imports, unnecessary files, unclear naming)

Let me know if you have any questions. Thanks!

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -1,41 +1,222 @@
...
+        help="If used, will load data via pandas",
+    )
+    parser.add_argument(
+        "-d",
+        "--download",
+        default=False,
+        action="store_true",
+        help="If used, will download from the source instead of uploading",
+    )
+    parser.add_argument(
```
Comment (ChengHaoKe) [inline run.py:150]: might be good to add in the readme that some of your cmd arguments are for the main argument parser and not for the subparsers and so the options should be added in front of the subparser names.

**Review by ChengHaoKe — APPROVED** [review]
Great work! 💯 
Love the github actions even if they are not vital, but it would be interesting to see if decreasing the number of javascript packages and css themes would still allow it to run, allowing this project to contain less files that are not central to the app itself.

I was able to run all of your functions inside docker 🤝

File: run.py [hunk 2]
```diff
@@ -1,41 +1,222 @@
...
+        "-s",
+        "--s3path",
+        default=DEFAULT_S3_LOCATION,
+        help="Location in S3 for source/destination file",
+    )
```
Comment (ChengHaoKe) [inline run.py:155]: might be good to add in the readme that some of your cmd arguments are for the main argument parser and not for the subparsers and so the options should be added in front of the subparser names.

File: requirements.txt [hunk 3]
```diff
@@ -1,6 +1,15 @@
...
-PyYAML==5.3.1
+boto3==1.12.32
+botocore==1.15.32
 Flask==1.1.1
+Flask-SQLAlchemy==2.4.1
+fsspec==0.8.4
+furo==2021.4.11b34
+matplotlib==3.4.1
+numpy==1.20.2
+pandas==1.2.4
```
[the remark this answers is not in the document]
Comment (brianrice2) [reply requirements.txt:9]: What do you mean?

File: run.py [hunk 4]
```diff
@@ -1,41 +1,222 @@
...
+        help="SQLAlchemy connection URI for database",
+    )
+    sp_ingest_album.add_argument("--album", default="Run the Jewels 2", help="Album title")
+    sp_ingest_album.add_argument("--artist", default="Run the Jewels", help="Artist of album")
+    sp_ingest_album.add_argument(
+        "--reviewauthor", default="Ian Cohen", help="Album reviewer's name"
+    )
+    sp_ingest_album.add_argument("--score", default="9", help="Pitchfork rating")
+    sp_ingest_album.add_argument("--releaseyear", default="2014", help="Album release year")
+    sp_ingest_album.add_argument(
```
[the remark this answers is not in the document]
Comment (brianrice2) [reply run.py:71]: Okay, I hesitate to intertwine the front and back ends in such a manner but will consider the implications 👍

File: config/flaskconfig.py [hunk 5]
```diff
@@ -1,26 +1,33 @@
...
 
 # Connection string
-DB_HOST = os.environ.get('MYSQL_HOST')
-DB_PORT = os.environ.get('MYSQL_PORT')
-DB_USER = os.environ.get('MYSQL_USER')
-DB_PW = os.environ.get('MYSQL_PASSWORD')
-DATABASE = os.environ.get('MYSQL_DATABASE')
-DB_DIALECT = 'mysql+pymysql'
-SQLALCHEMY_DATABASE_URI = os.environ.get('SQLALCHEMY_DATABASE_URI')
+DB_HOST = os.environ.get("MYSQL_HOST")
```
[the remark this answers is not in the document]
Comment (brianrice2) [reply config/flaskconfig.py:14]: Black auto-formatter 🤷🏻‍♂️

File: run.py (hunk 2 above)
Comment (brianrice2) [reply run.py]: Thanks for catching!

File: README.md [hunk 6]
```diff
@@ -4,27 +4,31 @@ Developer: Brian Rice
...
-  * [2. Configure Flask app](#2-configure-flask-app)
-  * [3. Run the Flask app](#3-run-the-flask-app)
-- [Running the app in Docker](#running-the-app-in-docker)
-  * [1. Build the image](#1-build-the-image)
-  * [2. Run the container](#2-run-the-container)
-  * [3. Kill the container](#3-kill-the-container)
-- [Testing](#testing)
+      - [RDS instance](#rds-instance)
+    - [Ingest the data](#ingest-the-data)
+  - [Testing](#testing)
```
[the remark this answers is not in the document]
Comment (brianrice2) [reply README.md:29]: Yes it is, and thanks 👍

**Comment by brianrice2** [comment]
I've incorporated your suggested changes. Thanks!

**Comment by ChengHaoKe** [comment]
Apparently i'm not very good at following instructions. Adding the checklist below:

## Functionality

Is the code doing what it’s meant to?  YEP

## Readability and code quality

    Is code easy to follow?  YES!
    Can you understand the design decisions that were made? In individual scripts as well as in the code base as a whole? YES
    Are variable and method names meaningful? Is it clear what they are referencing? YEP
    Are methods/functions appropriately sized? Or are there functions that need to be broken apart? Remember best practice is to decouple individual pieces of functionality (to enable testing and facilitate maintenance)  YES

## Overall design

    Does the code being reviewed make sense in the large context of the application? Is it consistent with the general approach that has so far been taken?  YES
    Is it modular? Is it extensible ?  YES
    Are there any issues with how the design might impact future development?  Added in my comments previously

## Documentation

    Are methods and functions properly documented?   YEP, additional comments added previously
    Thorough docstrings (Links to an external site.) (See link for best practices with Google formatting).
    Are inputs and outputs explained and referenced by the parameter name?  YEP
    Are there docstrings with descriptions of what functions/methods are doing?  YEP
    Is there commentary to aid others to understand what is being done throughout the code? (within functions, methods, etc)  YEP
    README.md (Links to an external site.)
        Does it explain the overall design / approach to the problem?  YES
        Does it explain how someone else can get the code, build it, and run it? I.e. are you able to clone their repository and get all functionality so far working without having to ask the developer questions?  YEP
    Are commit statements meaningful?  YES
    Are changes / additions explained in the PR itself?  YEP

**Pull request merged**